music.apple.com: some assets are not playing in Safari

jyavenard opened the pull request:
#### 5d82840740c54dca8ef43dccf2d716c0b79e5f49
<pre>
music.apple.com: some assets are not playing in Safari
<a href="https://bugs.webkit.org/show_bug.cgi?id=294067">https://bugs.webkit.org/show_bug.cgi?id=294067</a>
<a href="https://rdar.apple.com/152624427">rdar://152624427</a>

Reviewed by Youenn Fablet.

Some content when demuxed yield a few audio frames on start that have
a duration of 0 and invalid pts/dts. The underlying issue is being tracked
in <a href="https://rdar.apple.com/148824548">rdar://148824548</a>.
We were using the timescale of the presentation time to speedup time operations;
when the timestamp is invalid, that timescale will be 0. Attempting to reduce
a fraction with a zero denominator would have caused an infinite loop and
causing the GPU process to hang.

For now, we drop those samples.

Manually tested to avoid the issue. It is unknown on how the content was
created as it&apos;s not technically supposed to exist. The bad samples were
returned by the SourceBufferParserAVFObjC&apos;s AVStreamDataParser, as such
reproducing content exhibiting the problem is difficult.

* Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:
(WebCore::SourceBufferPrivate::processMediaSample): Handle the case where we
don&apos;t attempt to use the timescale of either the sample&apos;s presentationTime or
duration if they are invalid.
* Source/WebCore/platform/graphics/TrackBuffer.cpp:
(WebCore::roundTowardsTimeScaleWithRoundingMargin): Exit early if timescale provided is invalid
as it would have caused an infinite loop.

Canonical link: <a href="https://commits.webkit.org/295878@main">https://commits.webkit.org/295878@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/657bac80489355ccf09d19dc7a7bb5c4787364ba

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/106392 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/26141 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/16539 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/111593 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/56988 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/108431 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/26808 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/34644 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/80809 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/109396 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/21250 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/96013 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/61137 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/20734 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/14112 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/56428 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/90584 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/14147 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/114452 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/33530 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/24720 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/89881 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/33894 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/92242 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/89584 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22863 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/34472 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/12292 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/29133 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/33455 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/38867 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/33201 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/36554 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/34799 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:
```diff
@@ -855,6 +857,8 @@ bool SourceBufferPrivate::processMediaSample(SourceBufferPrivateClient& client,
             decodeTimestamp = sample->decodeTime();
         }
+        if (presentationTimestamp.isInvalid())
+            ERROR_LOG(LOGIDENTIFIER, "invalid sample time encountered:", sample.get());
```
ERROR_LOG_IF

youennf commented on Source/WebCore/platform/graphics/TrackBuffer.cpp:
```diff
@@ -42,6 +42,8 @@ static constexpr size_t MaximumSlidingWindowLength = 16;
 
 static inline MediaTime roundTowardsTimeScaleWithRoundingMargin(const MediaTime& time, uint32_t timeScale, const MediaTime& roundingMargin)
 {
+    if (!timeScale)
```
Should we `ASSERT(timeScale)` no that we checked the timeScale above?

jyavenard committed:
music.apple.com: some assets are not playing in Safari
https://bugs.webkit.org/show_bug.cgi?id=294067
rdar://152624427

Reviewed by Youenn Fablet.

Some content when demuxed yield a few audio frames on start that have
a duration of 0 and invalid pts/dts. The underlying issue is being tracked
in rdar://148824548.
We were using the timescale of the presentation time to speedup time operations;
when the timestamp is invalid, that timescale will be 0. Attempting to reduce
a fraction with a zero denominator would have caused an infinite loop and
causing the GPU process to hang.

For now, we drop those samples.

Manually tested to avoid the issue. It is unknown on how the content was
created as it's not technically supposed to exist. The bad samples were
returned by the SourceBufferParserAVFObjC's AVStreamDataParser, as such
reproducing content exhibiting the problem is difficult.

* Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:
(WebCore::SourceBufferPrivate::processMediaSample): Handle the case where we
don't attempt to use the timescale of either the sample's presentationTime or
duration if they are invalid.
* Source/WebCore/platform/graphics/TrackBuffer.cpp:
(WebCore::roundTowardsTimeScaleWithRoundingMargin): Exit early if timescale provided is invalid
as it would have caused an infinite loop.

Canonical link: https://commits.webkit.org/295878@main
